ff/handlers/nonbonded.py

```diff
@@ -161,6 +162,89 @@ def parameterize_ligand(params, param_idxs):
     return params[param_idxs]
 
 
+def compute_or_load_am1_charges(mol):
+    """Unless already cached in mol's "AM1Cache" property, use OpenEye to compute AM1 partial charges."""
+
+    # check for cache
+    cache_key = "AM1Cache"
+    if not mol.HasProp(cache_key):
+        # The charges returned by OEQuacPac is not deterministic across OS platforms. It is known
+        # to be an issue that the atom ordering modifies the return values as well. A follow up
+        # with OpenEye is in order
+        # https://github.com/openforcefield/openff-toolkit/issues/983
+        am1_charges = list(oe_assign_charges(mol, "AM1"))
+
+        mol.SetProp(cache_key, base64.b64encode(pickle.dumps(am1_charges)))
+
+    else:
+        am1_charges = pickle.loads(base64.b64decode(mol.GetProp(cache_key)))
+
+    return np.array(am1_charges)
+
+
+def bond_smirks_matches(mol, smirks_list):
+    """Return an array of ordered bonds and an array of their assigned types
+
+    Notes
+    -----
+    * Uses OpenEye for substructure searches
+    * Order within smirks_list matters
+        "First match wins."
+        For example, if bond (a,b) can be matched by smirks_list[2], smirks_list[5], ..., assign type 2
+    * Order within each smirks pattern matters
+        For example, "[#6:1]~[#1:2]" and "[#1:1]~[#6:2]" will match atom pairs in the opposite order
+    """
+    oemol = convert_to_oe(mol)
+    AromaticityModel.assign(oemol)
+
+    bond_idxs = []  # [B, 2]
+    type_idxs = []  # [B]
+
+    for type_idx, smirks in enumerate(smirks_list):
+        matches = oe_match_smirks(smirks, oemol)
+
+        for matched_indices in matches:
+            a, b = matched_indices[0], matched_indices[1]
+            forward_matched_bond = [a, b]
+            reverse_matched_bond = [b, a]
+
+            already_assigned = forward_matched_bond in bond_idxs or reverse_matched_bond in bond_idxs
+
+            if not already_assigned:
+                bond_idxs.append(forward_matched_bond)
+                type_idxs.append(type_idx)
+
+    return np.array(bond_idxs), np.array(type_idxs)
+
+
+def apply_bond_charge_corrections(initial_charges, bond_idxs, deltas):
+    """For an arbitrary collection of ordered bonds and associated increments `(a, b, delta)`,
+    update `charges` by `charges[a] += delta`, `charges[b] -= delta`
+
+    Notes
+    -----
+    * preserves sum(initial_charges) for arbitrary values of bond_idxs or deltas
```

Comment:
@proteneer : Could you double-check if there are any constraints on the inputs to this function?

(An antisymmetry constraint might have been suggested. Something along the lines of "for each bond `(a, b)`, we must provide either `(a, b, +delta)` or `(b, a, -delta)`.")

Reply:
Should just sanity check with assertions that initial charges and final charges:

1) sum up to an integer
2) sum up to the same integer

Reply:
Maybe also sanity check that bond_idxs are unique? Though that probably doesn't affect correctness but more likely an indication of user error...

Reply:
Charge sum preservation is logically guaranteed and was tested with randomized inputs (including duplicates, etc.) here https://github.com/proteneer/timemachine/blob/71734541fa88054b8b121ad47cfa97bc5b561213/tests/test_handlers.py#L621-L641

Runtime assertions added in https://github.com/proteneer/timemachine/pull/571/commits/4984f6d5a08b961a35d558f3c5b05ab90b8790b4 , randomized tests restricted to integral net charge in https://github.com/proteneer/timemachine/pull/571/commits/00bc62871d3480ba22265ead71c731256f85b9ff , and warnings for duplicate bond_idxs added in https://github.com/proteneer/timemachine/pull/571/commits/26ed49bcff38c187f1d6f46fb6f0699dce5523d9 .

Reply:
https://github.com/proteneer/timemachine/pull/571#discussion_r791309783

Ahh, shoot -- should have read this as "integer multiple of sqrt(ONE_4PI_EPS0)" not "integer", as @badisa identified and fixed in https://github.com/proteneer/timemachine/pull/680